### PR TITLE
Fix issue "Add quality check and cleanup for problematic unicode characters"

### LINKED ISSUE #10506: Add quality check and cleanup for problematic unicode characters
**Is your suggestion for improvement related to a problem? Please describe.**

Some unicode characters make problems, even with biblatex support (eg pdflatex still not completely supporting unicode). For example, `Garcı́a` gives
```
Package inputenc Error: Unicode character ́ (U+0301)
```

A few of such problematic characters are:
- U+0300: https://tex.stackexchange.com/questions/555086/package-inputenc-error-unicode-char-%cc%80-u300-inputenc-not-set-up-for-use-wit
- U+0301: see https://tex.stackexchange.com/questions/329239/how-to-solve-a-unicode-char-u301-error and https://tex.stackexchange.com/questions/443018/package-inputenc-error-unicode-char-%cc%81-u301inputenc?noredirect=1&lq=1
- U+2212: https://tex.stackexchange.com/questions/361019/unicode-error-in-inputenc-package

**Describe the solution you'd like**

As these characters are hard to recognize, it would be nice if there would be an integrity check warning about them, and an automatic cleanup to convert them to their unproblematic equivalents (e.g. [0131 + 0301](https://apps.timwhitlock.info/unicode/inspect?s=%C4%B1%CC%81) to [00ED](https://apps.timwhitlock.info/unicode/inspect?s=%C3%AD)).

**Additional context**
Might be helpful: https://github.com/zepinglee/citeproc-lua/blob/ab3ce712cc92073f12be26ff0b22b30eb906092d/citeproc/citeproc-latex-data.lua#L517

### PR DESCRIPTION
**Pull request by harsh1898**
fixes #10506 

- Added a new integrity check for problematic unicode fields that is not in NFC format. Now, it will show a message in the Quality/Integrity dialog box for non NFC values.
- Added a new cleanup feature in Quality/Cleanup Entries to format the problematic Unicode characters with Unicode Normalize formatter.
image
image
image
Here, we can see value is in NFC format after clean up.
 
### Mandatory checks

- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

File: src/main/java/org/jabref/logic/integrity/FieldCheckers.java [hunk 1]
```diff
@@ -53,6 +53,8 @@ private static Multimap<Field, ValueChecker> getAllMap(BibDatabaseContext databa
             fieldCheckers.put(StandardField.URLDATE, new DateChecker());
             fieldCheckers.put(StandardField.EVENTDATE, new DateChecker());
             fieldCheckers.put(StandardField.ORIGDATE, new DateChecker());
+//            fieldCheckers.put(StandardField.TITLE, new UnicodeNormalFormCCheck(databaseContext));
+//            fieldCheckers.put(StandardField.AUTHOR, new UnicodeNormalFormCCheck(databaseContext));
```
Comment (calixtus) [inline src/main/java/org/jabref/logic/integrity/FieldCheckers.java:57]: Why did you remove the feld checkers? Are they now included in your new Unicode Normal Form Checker?
If so, please remove them completly, otherwise we have a lot of dead code fragments.

Comment (harsh1898) [reply src/main/java/org/jabref/logic/integrity/FieldCheckers.java:57]: Hi @calixtus 
No, these particular filechecker that I removed , is not included in new Unicode Normal Form Checker.
Initially, I was trying here but it was not correct. So, I removed

Comment (calixtus) [reply src/main/java/org/jabref/logic/integrity/FieldCheckers.java:57]: But now they will remain commented. Will you uncomment these again or will you remove them?

Comment (harsh1898) [reply src/main/java/org/jabref/logic/integrity/FieldCheckers.java:57]: No, It will not uncomment again. 
We can completely remove them from here

Comment (calixtus) [reply src/main/java/org/jabref/logic/integrity/FieldCheckers.java:57]: ok, then go ahead, remove them and all the dead code related to them. 👍

Comment (harsh1898) [reply src/main/java/org/jabref/logic/integrity/FieldCheckers.java]: Hi @calixtus 

I have removed that unnecessary comment from particular FieldCheckers file.  Now, It is good

File: src/main/java/org/jabref/logic/formatter/bibtexfields/NormalizeUnicodeFormatter.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
+package org.jabref.logic.formatter.bibtexfields;
+
+import java.text.Normalizer;
+import java.util.Objects;
+
+import org.jabref.logic.cleanup.Formatter;
+
+/**
+ * Clean up field values by formatting Unicode values with Normalize Unicode
```
Comment (koppor) [inline src/main/java/org/jabref/logic/formatter/bibtexfields/NormalizeUnicodeFormatter.java:9]: Please provide the normal form reference there. The link should be in the issue description. Also use `html` here for linking.

File: src/main/java/org/jabref/logic/integrity/UnicodeNormalFormCCheck.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import org.jabref.logic.l10n.Localization;
+import org.jabref.model.entry.BibEntry;
+import org.jabref.model.entry.field.Field;
+
+/**
+ * Detect any Unicode characters that is not in NFC format
```
Comment (koppor) [inline src/main/java/org/jabref/logic/integrity/UnicodeNormalFormCCheck.java:13]: Please add a link to that format. (See other comment)

File: src/test/java/org/jabref/logic/integrity/UnicodeNormalFormCCheckTest.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+        assertEquals(Collections.emptyList(), checker.check(entry));
+    }
+
+    @Test
+    void checkWithNonNormalizedStringShouldReturnIntegrityMessage() {
+        entry.setField(StandardField.TITLE, "Café");
+        entry.setField(StandardField.AUTHOR, "John Doe");
+
+        assertFalse(checker.check(entry).isEmpty());
```
Comment (koppor) [inline src/test/java/org/jabref/logic/integrity/UnicodeNormalFormCCheckTest.java:31]: This line can be removed. The next statement is already covering that case: comparing a non-empty list with the result ensures that the result is non-empty.

File: src/test/java/org/jabref/logic/integrity/UnicodeNormalFormCCheckTest.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        entry.setField(StandardField.TITLE, "Some Title");
+        entry.setField(StandardField.AUTHOR, "John Doe");
+
+        assertEquals(Collections.emptyList(), checker.check(entry));
+    }
+
+    @Test
+    void checkWithNonNormalizedStringShouldReturnIntegrityMessage() {
+        entry.setField(StandardField.TITLE, "Café");
+        entry.setField(StandardField.AUTHOR, "John Doe");
```
Comment (koppor) [inline src/test/java/org/jabref/logic/integrity/UnicodeNormalFormCCheckTest.java:29]: Can you make `entry` a local variable and use `.withField` chains at the entry - and not `.setField`?

Also in the other test method.

File: src/main/java/org/jabref/logic/integrity/UnicodeNormalFormCCheck.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+ */
+public class UnicodeNormalFormCCheck implements EntryChecker {
+
+    @Override
+    public List<IntegrityMessage> check(BibEntry entry) {
+        List<IntegrityMessage> results = new ArrayList<>();
+        for (Map.Entry<Field, String> field : entry.getFieldMap().entrySet()) {
+            String normalizedString = Normalizer.normalize(field.getValue(), Normalizer.Form.NFC);
```
Comment (koppor) [inline src/main/java/org/jabref/logic/integrity/UnicodeNormalFormCCheck.java:21]: I checked the JavaDoc of normalizer now. It is available at https://docs.oracle.com/javase/6/docs/api/java/text/Normalizer.html.

Please use [isNormalized](https://docs.oracle.com/javase/6/docs/api/java/text/Normalizer.html#isNormalized(java.lang.CharSequence,%20java.text.Normalizer.Form)). That should be faster.

File: src/test/java/org/jabref/logic/integrity/UnicodeNormalFormCCheckTest.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    @Test
+    void checkWithNormalizedStringShouldReturnEmptyList() {
+        entry.setField(StandardField.TITLE, "Some Title");
+        entry.setField(StandardField.AUTHOR, "John Doe");
+
+        assertEquals(Collections.emptyList(), checker.check(entry));
+    }
+
+    @Test
+    void checkWithNonNormalizedStringShouldReturnIntegrityMessage() {
```
Comment (koppor) [inline src/test/java/org/jabref/logic/integrity/UnicodeNormalFormCCheckTest.java:27]: I have no clue on the normal forms. The [JavaDoc of the normalizer](https://docs.oracle.com/javase/6/docs/api/java/text/Normalizer.html#isNormalized(java.lang.CharSequence,%20java.text.Normalizer.Form)) gives other examples:

> Characters with accents or other adornments can be encoded in several different ways in Unicode. For example, take the character A-acute. In Unicode, this can be encoded as a single character (the "composed" form):
>
>      U+00C1    LATIN CAPITAL LETTER A WITH ACUTE
>
> or as two separate characters (the "decomposed" form):
>
>      U+0041    LATIN CAPITAL LETTER A
>      U+0301    COMBINING ACUTE ACCENT

Can you add a test for that, too? I assume that the first one is OK and the last one is NOT. Please use `@ParameterizedTest` and a `@MethodSource` to provide the input. Just pass the "title" as string to be set as title. The field "Author" can stay as is.

**Review by koppor — CHANGES_REQUESTED** [review]

**Comment by koppor** [comment]
@harsh1898 We did not see any comments or updates the last week. Are you still working on this?

**Comment by koppor** [comment]
I have no push permission. Closing and creating a new PR.
